CategoriesPanel.c

```diff
@@ -42,13 +42,13 @@ static void CategoriesPanel_delete(Object* object) {
 }
 
 static void CategoriesPanel_makeMetersPage(CategoriesPanel* this) {
-   size_t columns = HeaderLayout_getColumns(this->scr->header->headerLayout);
+   unsigned int columns = HeaderLayout_getColumns(this->scr->header->headerLayout);
    MetersPanel** meterPanels = xMallocArray(columns, sizeof(MetersPanel*));
    Settings* settings = this->host->settings;
 
-   for (size_t i = 0; i < columns; i++) {
+   for (unsigned int i = 0; i < columns; i++) {
       char titleBuffer[32];
-      xSnprintf(titleBuffer, sizeof(titleBuffer), "Column %zu", i + 1);
+      xSnprintf(titleBuffer, sizeof(titleBuffer), "Column %u", i + 1);
```

Comment:
I think this one is a step backwards. Stick witz `size_t`.

Reply:
Can't understand what you mean. What's the reason of keeping the `size_t`?

Reply:
Two-fold:

1. It's a number of items (columns)
2. `unsigned int` looks kinda unwieldy IMHO.
3. For the format string `%zu` is much more portable than `uint32_t` which would require you to use `%"PRIu32"` instead of just `%u`.

Reply:
Personal opinion following:

> @BenBE Well, I have a different view on this:
> 
> 1. Object sizes with no likelihood of overflowing an `int` type would not need `size_t` (that is, `unsigned long`) for storing.

Nonetheless `size_t` is less about the number of entities itself, but that you are **counting** entities.

Otherwise you could argue that "this hashtable implementation only needs `short` for its size, because I never store more than 65535 elements in it".

I know, that example is hyperbole, but follows that same logic. When I ask for `size_t` it's really little about "how many", but the "what" (AKA semantics).

> 2. In x86-64, incrementing and comparing `int` types are slightly cheaper than comparing `long` or `size_t` (in terms of code size). When `int` types can suffice, using `long` would make the code one-byte-per-operation larger.

That might be true for x86_64, but that's not the only platform we need to support. There's at least ARM, ARM64, MIPS, and probably several others, where htop is runing. Unless performance really IS a bottleneck for these functions (AFAICS it's not), priority should be on correctness, not code size or performance.

> 3. Header column count is unlikely to exceed terminal column count (`COLS`), and the latter is bound to `signed int` type already.

See arguments from the article you linked.

That limit is implied by the current interface from ncurses, but a future version could just switch to size_t for its column and row counts (potentially breaking BC at that point); there are good reasons like huge screens where terminals COULD have that many columns/rows.

> 4. And no, we don't need `PRIu32` here. It's `unsigned int`, not `uint32_t`.

k, ACK on that. But that's only a minor note regarding the chosen type signature.

> This case is slightly different from string lengths. My view is string lengths in `size_t` can serve as a defence in depth, where the user is possible to feed a 4 GiB large string in a config file. But if the object size would be bound by any other factor, then `size_t` might be a waste.

Allocations are typically backed by 4 KiB pages at minimum, with the usual memory manager implementation pre-allocating some memory to handle sudden spikes. Thus unless the savings are major or we have massive amounts of objects (where storing UTF-8 instead of UTF-16 saves several hundreds of megabytes (like with some browsers) there's usually little need to save every last byte. Furthermore, by storing smaller data types you sometimes may make the alignment of the data worse, thus actually reducing performance.

From my experience I can tell, that I've seen real-life cases of where `uint16_t` vs. `uint32_t` was actually harmful due to alignment and memory access patterns. For 32 bit values on a 64 bit architecture, similar considerations apply.

> (By the way, I did read [this article from EWONTFIX](http://ewontfix.com/9/) when I write my changes.)

FWIW, I personally side with the author.

> I hope htop developers can reconsider.

@cgzones @fasterit @ginggs @natoscott : Your opinion on this matter of `size_t` vs `unsigned int`?

Reply:
I'm happy to stick with size_t here.

Reply:
Seconded
/DLange

Reply:
@BenBE @natoscott @fasterit 
This commit was to fix the implicit downcast warnings from `HeaderLayout_getColumns()` by Clang 19 (the messages was: "`implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'unsigned int' [-Wshorten-64-to-32]`"). `HeaderLayout_getColumns()` was an inline function that returns a `size_t` that, in certain caller codes, has to downcast to `int` because terminal column variables are in `int` type to follow `curses` API.

With this background, I would like some clarifications on which decision to make:
(A) Keep `HeaderLayout_getColumns()` return value in `size_t`. Use `size_t` for local variables that hold the return values of `HeaderLayout_getColumns()`, and add explicit downcasts when needed (e.g. when used in the context of terminal column widths).
(B) Let `HeaderLayout_getColumns()` return an `unsigned int`. Keep using `int` type for places like terminal columns. But when using in array iterators (like in this example code you guys are commented on), upcast to `size_t`.
(C) Let `HeaderLayout_getColumns()` return an `unsigned int`. Also use `unsigned int` for array iterators that would never cross the bound of `HeaderLayout_getColumns()`. This is my original proposal.

Reply:
A.